### DISCUSSION TITLE
How to install python3 dirac client in a different location to where it will be used?

### DISCUSSION THREAD

**Discussion by marianne013** [DIRACGrid/DIRAC#6098] [Q&A] [answered]
When installing software in cvmfs I am required to install it (on the cvmfs stratum 0 node) in:
/cvmfs-mirror/dirac.egi.eu/dirac/v7r2p45_py3
^^^^^^^^^^^^^
/cvmfs does not exist.
I tried to install the diracos bundle naively, but ended up with /cvmfs-mirror hardcoded in a lot of files.
I then tried using the -p option, but that fails:
```
[diracsgm@cvmfs-uploader02 v7r2p45_py3]$ bash DIRACOS-Linux-x86_64.sh -p /cvmfs/dirac.egi.eu/dirac/v7r2p45_py3/diracos
mkdir: cannot create directory ‘/cvmfs’: Permission denied
ERROR: Could not create directory: '/cvmfs/dirac.egi.eu/dirac/v7r2p45_py3/diracos'
```
How do I convince the UI to put the correct path in all its scripts ?

**Comment by marianne013** [comment]
@chrisburr You ignoring me  ;-) ?

**Reply by andresailer** [reply]
You can't open a cvmfs transaction and install directly in the right location?

**Comment by marianne013** [comment, marked as answer]
Unfortunately that's not possible on the RAL stratum0. /cvmfs does not exist there, it's /cvmfs-mirror instead.

**Reply by chrisburr** [reply]
Installing in `/cvmfs-mirror` and running from `/cvmfs` won't work as the binaries aren't relocatable after they're installed. Does RAL really need to have such a weird setup?

Assuming nothing can be done about the Stratum 0. Your best bet might be to make a fake CVMFS tree to do the install (possibly inside a container) and then copy it to `/cvmfs-mirror/...`, i.e. something like:

```bash
FAKE_CVMFS=$(mktemp -d)
singularity exec --bind "$FAKE_CVMFS:/cvmfs" ....
bash DIRACOS-Linux-x86_64.sh -p /cvmfs/dirac.egi.eu/dirac/v7r2p45_py3/diracos
source /cvmfs/dirac.egi.eu/dirac/v7r2p45_py3/diracos/diracosrc
pip install DIRAC=7.2.45
...
exit

mkdir -p /cvmfs-mirror/dirac.egi.eu/dirac/v7r2p45_py3/diracos
cp -rp "$FAKE_CVMFS/dirac.egi.eu/dirac/v7r2p45_py3/diracos" /cvmfs-mirror/dirac.egi.eu/dirac/v7r2p45_py3/diracos
```

**Reply by chrisburr** [reply]
Actually there might be a smarter workaround:

```bash
singularity exec --bind "/cvmfs-mirror:/cvmfs" bash my_install_script.sh
```

**Reply by fstagni** [reply]
@marianne013 was this idea tried out?

**Reply by marianne013** [reply]
It's on my to do list. We need a "on hold" feature for these things.
RAL has adjusted their cvmfs, so that took the pressure off a bit.

**Reply by marianne013** [reply]
For the record this seems to work (at Imperial, using a native install of singularity):
mkdir tmp
cd tmp
singularity exec --bind "/home/hep/dbauer/tmp:/cvmfs" /cvmfs/cernvm-prod.cern.ch/cvm4/ /bin/bash

curl -LO https://github.com/DIRACGrid/DIRACOS2/releases/latest/download/DIRACOS-Linux-$(uname -m).sh
bash DIRACOS-Linux-x86_64.sh -p /cvmfs/dirac.egi.eu/dirac/v7r2p45_py3/diracos
source /cvmfs/dirac.egi.eu/dirac/v7r2p45_py3/diracos/diracosrc
pip install DIRAC==7.2.45

However RAL just having /cvmfs mounted on the actual upload nodes is a lot less error prone, so this remains our favourite solution, especially as the upload nodes don't seem to have singularity installed.

I've selected an answer, so this will hopefully close the issue.

**Comment by marianne013** [comment]
https://ggus.eu/?mode=ticket_info&ticket_id=157707

**Comment by fstagni** [comment]
Am I right if I say that this just turns out to be an issue with the RAL CVMFS configuration and that has nothing to do with DIRAC? If that's the case, please select an "answer" so we close this discussion.

**Comment by marianne013** [comment]
@fstagni Depends, the python2 UI was fully relocatable and could be installed directly at RAL (which for the better or worse hosts /cvmfs/dirac.egi.eu) and python 3 isn't and that is certainly something for the admins to be aware of (as in "it's fine if you are LHCb, but ....") . @sfayer is still trying to convince RAL to see the error of their ways.

**Reply by chrisburr** [reply]
> the python2 UI was fully relocatable

This is something that was explicitly stated as a feature that would be removed as part of the Python 3 transition (as it's painful to support and didn't even work properly with DIRACOS1). Regardless of the outcome at RAL the solution to installing in a non-runtime location is in https://github.com/DIRACGrid/DIRAC/discussions/6098#discussioncomment-2819174.
